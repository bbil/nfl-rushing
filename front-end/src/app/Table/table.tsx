import React from 'react';
import { connect } from 'react-redux';

import { getData } from '../../redux/rushing/selectors';
import { initialFetch } from '../../redux/rushing/thunk';
import { NflRushingDataItem } from '../../redux/rushing/types';
import { State } from '../../redux/store';
import { Pagination } from './pagination';

interface TableDispatchProps {
    fetchData: () => void;
}

interface TableStateProps {
    nflRushingData: NflRushingDataItem[];
}

export type TableProps = TableDispatchProps & TableStateProps;

export const TableComponent: React.FC<TableProps> = props => {
    // on mount; load first-time data
    React.useEffect(() => {
        props.fetchData();
    }, []);

    return (
        <>
            <table className="table table-bordered table-striped">
                <TableHead />
                <TableBody data={props.nflRushingData} />
            </table>
            <Pagination />
        </>
    );
};

function mapDispatchToProps(dispatch: Function): TableDispatchProps {
    return {
        fetchData: () => dispatch(initialFetch())
    }
}

function mapStateToProps(state: State): TableStateProps {
    return {
        nflRushingData: getData(state)
    }
}

export const Table = connect(mapStateToProps, mapDispatchToProps)(TableComponent);

export const TableHead: React.FC = () => (
  <thead>
    <tr>
      <th>Name</th>
      <th>Team</th>
      <th>Position</th>
      <th>Rushing Attempts</th>
      <th>Rushing Attempts / Game</th>
      <th>Total Rushing Yards</th>
      <th>Rushing Yards / Game</th>
      <th>Rushing Yards / Attempt</th>
      <th>Rushing Touchdowns</th>
      <th>Longest Rush</th>
      <th>Longest Rush was a touchdown</th>
      <th>Rush First Downs</th>
      <th>Rush First Downs %</th>
      <th>Rush 20+ Yards</th>
      <th>Rush 40+ Yards</th>
      <th>Rush Fumbles</th>
    </tr>
  </thead>
);

export type TableBodyProps = { data: NflRushingDataItem[] };
export const TableBody: React.FC<TableBodyProps> = ({ data }) => (
  <tbody>
    {data.map(player => {
        return (
            <TableRow key={player.name} player={player} />
        )
    })}
  </tbody>
);

type TableRowProps = { player: NflRushingDataItem };
const TableRow: React.FC<TableRowProps> = ({ player }) => (
    <tr>
        <td>{player.name}</td>
        <td>{player.team}</td>
        <td>{player.position }</td>
        <td>{player.rushingAttempts}</td>
        <td>{player.rushingAttemptsPerGame }</td>
        <td>{player.totalRushingYards}</td>
        <td>{player.rushingYardsPerAttempt }</td>
        <td>{player.rushingYardsPerGame }</td>
        <td>{player.totalRushingTouchdowns}</td>
        <td>{player.longestRush}</td>
        <td>{player.longestRushTouchdown ? <>&#10003;</> : <>&#10007;</>}</td>
        <td>{player.rushFirstDowns}</td>
        <td>{player.rushFirstDownPercent}</td>
        <td>{player.rush20Plus}</td>
        <td>{player.rush40Plus}</td>
        <td>{player.rushFumbles}</td>
    </tr>
);
