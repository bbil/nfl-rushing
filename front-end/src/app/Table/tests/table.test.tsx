import React from 'react';
import '@testing-library/react';

import { shallow } from 'enzyme';

import { TableComponent, TableProps } from '../table';

describe('Table Component', () => {
    const PROPS: TableProps = {
        fetchData: jest.fn(),
        nflRushingData: []
    };

    beforeEach(() => {
        // have useEffect actually run
        jest.spyOn(React, 'useEffect').mockImplementation(f => f());
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('renders', () => {
        const wrapper = shallow(<TableComponent {...PROPS} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('calls initial on mount', () => {
        shallow(<TableComponent {...PROPS} />);

        expect(PROPS.fetchData).toHaveBeenCalled();
    });
});
