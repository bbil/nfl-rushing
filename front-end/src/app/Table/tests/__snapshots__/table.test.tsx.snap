// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component renders 1`] = `
<Fragment>
  <table
    className="table table-bordered table-striped"
  >
    <TableHead />
    <TableBody
      data={Array []}
    />
  </table>
  <Connect(PaginationComponent) />
</Fragment>
`;
